### PR TITLE
Dns response 2198 v7

### PR DESCRIPTION
**Pull request by StianHB**
- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues/) ticket:[2198](https://redmine.openinfosecfoundation.org/issues/2198)

Describe changes:
- Created the dns_response keyword that allows matching of byte strings on dns responses
- Updated the user-docs to reflect the new keyword

Incorporated changes from #3004, #3006, #3057, #3061 and #3111

### CODE REVIEW

**Comment by victorjulien** [comment]
<pre>
==28544== ERROR SUMMARY: 6 errors from 1 contexts (suppressed: 0 from 0)
==28544== 
==28544== 6 errors in context 1 of 1:
==28544== Invalid read of size 1
==28544==    at 0x8B027A: SCACSearch (util-mpm-ac.c:1048)
==28544==    by 0x5E05D8: PrefilterTxDnsResponse (detect-engine-dns.c:265)
==28544==    by 0x659A9F: PrefilterTx (detect-engine-prefilter.c:138)
==28544==    by 0x659C7D: Prefilter (detect-engine-prefilter.c:206)
==28544==    by 0x557165: SigMatchSignatures (detect.c:742)
==28544==    by 0x599FD8: DetectDnsResponseTest01 (detect-dns-response.c:192)
==28544==    by 0x902441: UtRunTests (util-unittest.c:214)
==28544==    by 0x7EDF9C: RunUnittests (runmode-unittests.c:301)
==28544==    by 0x8533BC: StartInternalRunMode (suricata.c:2300)
==28544==    by 0x8551F6: main (suricata.c:2802)
==28544==  Address 0x7c9ac1e is 0 bytes after a block of size 46 alloc'd
==28544==    at 0x4C28C20: malloc (vg_replace_malloc.c:296)
==28544==    by 0x482805: DNSStoreAnswerInState (app-layer-dns-common.c:611)
==28544==    by 0x48315A: DNSReponseParse (app-layer-dns-common.c:856)
==28544==    by 0x4864B4: DNSUDPResponseParse (app-layer-dns-udp.c:282)
==28544==    by 0x4B29B1: AppLayerParserParse (app-layer-parser.c:1051)
==28544==    by 0x599F21: DetectDnsResponseTest01 (detect-dns-response.c:181)
==28544==    by 0x902441: UtRunTests (util-unittest.c:214)
==28544==    by 0x7EDF9C: RunUnittests (runmode-unittests.c:301)
==28544==    by 0x8533BC: StartInternalRunMode (suricata.c:2300)
==28544==    by 0x8551F6: main (suricata.c:2802)
==28544== 
==28544== ERROR SUMMARY: 6 errors from 1 contexts (suppressed: 0 from 0)
</pre>

**Comment by victorjulien** [comment]
ASAN sees it too:
<pre>
Test DetectDnsResponseTest01                                      : [100123] 9/1/2018 -- 12:05:43 - (util-classification-config.c:370) <Info> (SCClassConfParseFile) -- Added "34" classification types from the classification file
[100123] 9/1/2018 -- 12:05:43 - (util-reference-config.c:352) <Info> (SCRConfParseFile) -- Added "19" reference types from the reference.config file
=================================================================
==91391==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6040000001be at pc 0x00000185e2a6 bp 0x7ffe07bce2b0 sp 0x7ffe07bce2a8
READ of size 1 at 0x6040000001be thread T16777215
</pre>

File: src/detect-engine-dns.c [hunk 1]
```diff
@@ -119,6 +119,65 @@ int DetectEngineInspectDnsQueryName(ThreadVars *tv,
...
+    TAILQ_FOREACH(answer, &tx->answer_list, next) {
+        SCLogDebug("tx %p response %p", tx, answer);
+
+        det_ctx->discontinue_matching = 0;
+        det_ctx->buffer_offset = 0;
+        det_ctx->inspection_recursion_counter = 0;
+
+        buffer = (uint8_t *) ((uint8_t *) answer +
+                    sizeof(DNSAnswerEntry) + answer->fqdn_len);
+        buffer_len = answer->fqdn_len;
```
Comment (victorjulien) [inline src/detect-engine-dns.c]: I think the actual data len is 'answer->data_len'. See the logging code, for example https://github.com/OISF/suricata/blob/master/src/output-json-dns.c#L509

Comment (StianHB) [reply src/detect-engine-dns.c]: Yeah, that makes sense. Ill submit a new PR.

**Pull request closed without merging**